File: pxr/usdImaging/usdNoodles/nodeLibs/usdPrimRegistry.py
#
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# Licensed under the terms set forth in the LICENSE.txt file available
# at the root of this repository.
#

#
# pyre-strict

"""
Registry of concrete USD prim types discovered from UsdSchemaRegistry.

Pure-Python replacement for the former C++ ``noodles::UsdPrimRegistry``. It
discovers every concrete typed prim schema known to ``UsdSchemaRegistry`` and
groups them by their plugin family (``usdGeom``, ``usdLux``, ...) so the create
menu can offer every concrete type rather than a fixed allowlist.

Discovery runs once on first access and is cached on the instance.

Attribute introspection and connection reading (``GetDescriptorFromPrim``,
``GetLinksForPrim``, ``GetSchemaAwarePropertyNames``,
``GetAttributeTypeNameString`` on the old C++ class) now live in
``_schema_pin_names.py`` and ``pinUtils.py`` and are the single source of truth
for pin discovery, so only schema-type discovery remains here.
"""

from __future__ import annotations

from dataclasses import dataclass

from pxr import Plug, Tf, Usd


@dataclass(frozen=True)
class PrimTypeInfo:
    """A concrete USD prim type and the plugin family it belongs to.

    Field names mirror the former C++ ``UsdPrimRegistry::PrimTypeInfo`` struct
    so existing callers reading ``.typeName`` / ``.family`` are unchanged.
    """

    typeName: str
    family: str


class UsdPrimRegistry:
    """Discovers concrete USD prim types from ``UsdSchemaRegistry``.

    Types are grouped by plugin family and sorted alphabetically within each
    family. Discovery is performed once on first access and cached on the
    instance.

    Not thread-safe: construct and query from a single thread (the Qt main
    thread), matching USD's own ``Plug.Registry`` / ``Usd.SchemaRegistry``,
    which are also not thread-safe for the queries used here.
    """

    def __init__(self) -> None:
        # None until _discover_types() runs; acts as the once-only guard.
        self._all_types: list[PrimTypeInfo] | None = None
        self._family_index: dict[str, list[int]] = {}

    def _discover_types(self) -> None:
        if self._all_types is not None:
            return

        all_types: list[PrimTypeInfo] = []
        family_index: dict[str, list[int]] = {}

        base_type = Tf.Type.FindByName("UsdSchemaBase")
        plug_registry = Plug.Registry()

        for tf_type in base_type.GetAllDerivedTypes():
            try:
                # IsConcrete excludes abstract and API schemas. When a schema
                # type's plugin is not registered in the current environment,
                # USD posts a Tf error (surfaced in Python as Tf.ErrorException)
                # while resolving its schema kind; skip such types instead of
                # aborting, mirroring the former C++ code, where the same
                # condition simply yielded a non-concrete result.
                if not Usd.SchemaRegistry.IsConcrete(tf_type):
                    continue

                type_name = str(Usd.SchemaRegistry.GetSchemaTypeName(tf_type))
                if not type_name:
                    continue

                # Family from the plugin name, reliable across USD versions.
                plugin = plug_registry.GetPluginForType(tf_type)
                family = plugin.name if plugin is not None else "Other"
            except Tf.ErrorException:
                continue

            index = len(all_types)
            all_types.append(PrimTypeInfo(typeName=type_name, family=family))
            family_index.setdefault(family, []).append(index)

        # Sort indices within each family alphabetically by type name.
        for indices in family_index.values():
            indices.sort(key=lambda i: all_types[i].typeName)

        self._all_types = all_types
        self._family_index = family_index

    def GetFamilies(self) -> list[str]:
        """Return sorted plugin family names (e.g. ``usdGeom``)."""
        self._discover_types()
        return sorted(self._family_index)

    def GetPrimTypes(self, family: str) -> list[PrimTypeInfo]:
        """Return concrete prim types in ``family``, sorted by type name."""
        self._discover_types()
        indices = self._family_index.get(family)
        if not indices:
            return []
        all_types = self._all_types or []
        return [all_types[index] for index in indices]

File: pxr/usdImaging/usdNoodles/usd/UsdPrimRegistry.cpp
//
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// Licensed under the terms set forth in the LICENSE.txt file available
// at the root of this repository.
//

#include "usd/UsdPrimRegistry.h"

#include <pxr/base/plug/plugin.h>
#include <pxr/base/tf/type.h>
#include <pxr/usd/sdf/attributeSpec.h>
#include <pxr/usd/sdf/valueTypeName.h>
#include <pxr/usd/usd/attribute.h>
#include <pxr/usd/usd/primDefinition.h>
#include <pxr/usd/usd/schemaBase.h>
#include <pxr/usd/usd/schemaRegistry.h>

#include <algorithm>
#include <set>
#include <string_view>
#include <tuple>

namespace {
constexpr std::string_view kInputsPrefix = "inputs:";
constexpr std::string_view kOutputsPrefix = "outputs:";
} // namespace

// ---------------------------------------------------------------------------
// Schema discovery
// ---------------------------------------------------------------------------

void UsdPrimRegistry::_DiscoverTypes() const {
  std::call_once(_discoveryFlag, [this]() {
    const auto& schemaReg = UsdSchemaRegistry::GetInstance();
    std::set<TfType> allDerived;
    TfType::FindByName("UsdSchemaBase").GetAllDerivedTypes(&allDerived);

    auto& plugReg = PlugRegistry::GetInstance();

    // Discover every concrete schema type, grouped by its plugin family.
    // The generic UsdPrimLibrary renders all prim types now, so the create
    // menu offers every concrete type (abstract + API schemas are still
    // excluded by the IsConcrete check below) rather than a fixed allowlist.

    for (const auto& type : allDerived) {
      if (!schemaReg.IsConcrete(type)) {
        continue; // Excludes abstract + API schemas
      }

      std::string typeName = schemaReg.GetSchemaTypeName(type);
      if (typeName.empty()) {
        continue;
      }

      // Get family from plugin name (reliable across USD versions)
      std::string family = "Other";
      if (auto plugin = plugReg.GetPluginForType(type)) {
        family = plugin->GetName();
      }

      size_t idx = _allTypes.size();
      _allTypes.push_back({typeName, family});
      _familyIndex[family].push_back(idx);
      _typeNameSet.insert(typeName);
    }

    // Sort indices within each family alphabetically by typeName
    for (auto& [fam, indices] : _familyIndex) {
      std::sort(indices.begin(), indices.end(), [this](size_t a, size_t b) {
        return _allTypes[a].typeName < _allTypes[b].typeName;
      });
    }
  });
}

std::vector<std::string> UsdPrimRegistry::GetFamilies() const {
  _DiscoverTypes();

  std::vector<std::string> families;
  families.reserve(_familyIndex.size());
  for (const auto& [fam, _] : _familyIndex) {
    families.push_back(fam);
  }
  std::sort(families.begin(), families.end());
  return families;
}

std::vector<UsdPrimRegistry::PrimTypeInfo> UsdPrimRegistry::GetPrimTypes(
    const std::string& family) const {
  _DiscoverTypes();

  auto it = _familyIndex.find(family);
  if (it == _familyIndex.end()) {
    return {};
  }

  std::vector<PrimTypeInfo> result;
  result.reserve(it->second.size());
  for (size_t idx : it->second) {
    result.push_back(_allTypes[idx]);
  }
  return result;
}

// ---------------------------------------------------------------------------
// Attribute introspection
// ---------------------------------------------------------------------------

UsdPrimRegistry::PrimDescriptor UsdPrimRegistry::GetDescriptorFromPrim(const UsdPrim& prim) const {
  PrimDescriptor desc;
  desc.primPath = prim.GetPath().GetString();

  std::unordered_set<std::string> inputSeen, outputSeen;
  for (const TfToken& nameTok : GetSchemaAwarePropertyNames(prim)) {
    const std::string& name = nameTok.GetString();
    const std::string typeName = GetAttributeTypeNameString(prim, nameTok);

    if (name.rfind(kInputsPrefix, 0) == 0) {
      std::string pinName = name.substr(kInputsPrefix.size());
      if (inputSeen.insert(pinName).second) {
        desc.inputPins.push_back(pinName);
      }
      if (!typeName.empty()) {
        desc.inputPinTypes[pinName] = typeName;
      }
    } else if (name.rfind(kOutputsPrefix, 0) == 0) {
      std::string pinName = name.substr(kOutputsPrefix.size());
      if (outputSeen.insert(pinName).second) {
        desc.outputPins.push_back(pinName);
      }
      if (!typeName.empty()) {
        desc.outputPinTypes[pinName] = typeName;
      }
    } else if (name.find(':') == std::string::npos) {
      if (inputSeen.insert(name).second) {
        desc.inputPins.push_back(name);
      }
      if (outputSeen.insert(name).second) {
        desc.outputPins.push_back(name);
      }
      if (!typeName.empty()) {
        desc.inputPinTypes[name] = typeName;
        desc.outputPinTypes[name] = typeName;
      }
    }
  }
  return desc;
}

// ---------------------------------------------------------------------------
// Schema-aware property helpers
// ---------------------------------------------------------------------------

TfTokenVector UsdPrimRegistry::GetSchemaAwarePropertyNames(const UsdPrim& prim) {
  // TfToken dedup: its HashFunctor hashes the interned pointer, avoiding a
  // std::string allocation per token compared to std::unordered_set<string>.
  std::unordered_set<TfToken, TfToken::HashFunctor> seen;
  TfTokenVector ordered;

  // UsdPrim::GetPrimDefinition() returns the per-prim composed definition,
  // which includes concrete type attrs, built-in API schemas, and any
  // runtime-applied API schemas on this specific prim.
  const UsdPrimDefinition& primDef = prim.GetPrimDefinition();
  for (const TfToken& name : primDef.GetPropertyNames()) {
    // Skip relationships — only attributes become pins.
    if (!primDef.GetSchemaAttributeSpec(name)) {
      continue;
    }
    if (seen.insert(name).second) {
      ordered.push_back(name);
    }
  }

  for (const UsdAttribute& attr : prim.GetAuthoredAttributes()) {
    const TfToken& name = attr.GetName();
    if (seen.insert(name).second) {
      ordered.push_back(name);
    }
  }
  return ordered;
}

std::string UsdPrimRegistry::GetAttributeTypeNameString(
    const UsdPrim& prim,
    const TfToken& attrName) {
  if (UsdAttribute attr = prim.GetAttribute(attrName); attr && attr.IsValid()) {
    if (SdfValueTypeName t = attr.GetTypeName()) {
      return t.GetAsToken().GetString();
    }
  }

  const UsdPrimDefinition& primDef = prim.GetPrimDefinition();
  if (SdfAttributeSpecHandle spec = primDef.GetSchemaAttributeSpec(attrName); spec) {
    return spec->GetTypeName().GetAsToken().GetString();
  }
  return {};
}

// ---------------------------------------------------------------------------
// Prim type check
// ---------------------------------------------------------------------------

bool UsdPrimRegistry::CanHandlePrim(const UsdPrim& prim) const {
  _DiscoverTypes();

  std::string typeName = prim.GetTypeName();
  if (typeName.empty()) {
    return false;
  }

  return _typeNameSet.count(typeName) > 0;
}

// ---------------------------------------------------------------------------
// Connection reading
// ---------------------------------------------------------------------------

std::vector<UsdPrimRegistry::LinkInfo> UsdPrimRegistry::GetLinksForPrim(const UsdPrim& prim) const {
  std::vector<LinkInfo> links;
  // Bare attributes can produce duplicates when both endpoints are bare
  // and the connection is discovered from both sides. Dedup inline so we
  // never carry duplicates in `links` to begin with.
  std::set<std::tuple<std::string, std::string, std::string, std::string>> seen;

  for (const auto& attr : prim.GetAttributes()) {
    std::string attrName = attr.GetName();

    bool isInput = false;
    std::string pinName;

    bool isBare = false;
    if (attrName.rfind(kInputsPrefix, 0) == 0) {
      isInput = true;
      pinName = attrName.substr(kInputsPrefix.size());
    } else if (attrName.rfind(kOutputsPrefix, 0) == 0) {
      isInput = false;
      pinName = attrName.substr(kOutputsPrefix.size());
    } else if (attrName.find(':') == std::string::npos) {
      // Bare attribute - dual pin, direction inferred per-connection
      isBare = true;
      isInput = true; // placeholder
      pinName = attrName;
    } else {
      continue;
    }

    SdfPathVector connections;
    attr.GetConnections(&connections);

    for (const auto& connPath : connections) {
      // Use SdfPath API to split into prim path and property name
      if (!connPath.IsPrimPropertyPath()) {
        continue;
      }

      std::string targetPrimPath = connPath.GetPrimPath().GetString();
      std::string targetPropName = connPath.GetNameToken().GetString();

      // Extract target pin name
      std::string targetPinName;
      if (targetPropName.rfind(kInputsPrefix, 0) == 0) {
        targetPinName = targetPropName.substr(kInputsPrefix.size());
      } else if (targetPropName.rfind(kOutputsPrefix, 0) == 0) {
        targetPinName = targetPropName.substr(kOutputsPrefix.size());
      } else {
        targetPinName = targetPropName;
      }

      // For bare attributes, infer direction from target property prefix.
      // Target inputs:* -> bare is the output (source) side.
      // Target outputs:* or bare -> bare is the input (target) side
      // (UsdShade convention).
      if (isBare) {
        isInput = targetPropName.rfind(kInputsPrefix, 0) != 0;
      }

      LinkInfo link;
      if (isInput) {
        // Connection on input side: this node is target, connection points to
        // source
        link.sourceNodeId = targetPrimPath;
        link.sourcePort = targetPinName;
        link.targetNodeId = prim.GetPath().GetString();
        link.targetPort = pinName;
        link.isInputLink = true;
      } else {
        // Connection on output side: this node is source, connection points to
        // target
        link.sourceNodeId = prim.GetPath().GetString();
        link.sourcePort = pinName;
        link.targetNodeId = targetPrimPath;
        link.targetPort = targetPinName;
        link.isInputLink = false;
      }

      auto key =
          std::make_tuple(link.sourceNodeId, link.sourcePort, link.targetNodeId, link.targetPort);
      if (seen.insert(key).second) {
        links.push_back(link);
      }
    }
  }

  return links;
}
